core-to-plc/src/Language/Plutus/Lift/LiftPir.hs

```diff
@@ -0,0 +1,349 @@
+    pure [|
+          -- we won't know the result type until runtime, so take it as an argument
+          \resultType -> do
+              tys' <- sequence $(TH.listE tyExprs)
+              let constrTy = mkIterTyFun () tys' resultType
+              constrName <- safeFreshName () $ showName name
+              pure $ VarDecl () constrName constrTy
+          |]
+
+makeTypeablePir :: TH.Name -> TH.Q [TH.Dec]
+makeTypeablePir name = do
+    info <- TH.reifyDatatype name
+    let (rhs, deps) = runState (compileTypeRep info) mempty
+
+    -- See Note [Closed constraints]
+    let constraints = filter (not . isClosedConstraint) $ toConstraint <$> Set.toList deps
+
+    decl <- TH.funD 'typeRep [TH.clause [TH.wildP] (TH.normalB rhs) []]
+    pure [TH.InstanceD Nothing constraints (typeablePir (TH.ConT name)) [decl]]
+
+-- LiftPir
+
+-- | Class for types which can be lifted into Plutus IR.
+class LiftPir a where
+    -- | Get a Plutus IR term corresponding to the given value.
+    lift :: (RTCompiling m) => a -> m (Term TyName Name ())
+
+compileLift :: THCompiling m => TH.DatatypeInfo -> m [TH.Q TH.Clause]
+compileLift dt = traverse (uncurry (compileConstructorClause dt)) (zip [0..] (sortedCons dt))
+
+compileConstructorClause :: (THCompiling m) => TH.DatatypeInfo -> Int -> TH.ConstructorInfo -> m (TH.Q TH.Clause)
+compileConstructorClause TH.DatatypeInfo{TH.datatypeName=tyName, TH.datatypeVars=tvs} index TH.ConstructorInfo{TH.constructorName=name, TH.constructorFields=argTys} = do
+    -- need to be able to lift the argument types
+    traverse_ addLiftDep argTys
+
+    -- need the actual type parameters
+    typeExprs <- traverse (compileType Typeable . normalizeType) tvs
+    pure $ do
+        patNames <- for argTys $ \_ -> TH.newName "arg"
+        let pats = fmap TH.varP patNames
+        let pat = TH.conP name pats
+        let liftExprs = fmap (\pn -> TH.varE 'lift `TH.appE` TH.varE pn) patNames
+        -- see note [Compiling at TH time and runtime]
+        expr <-
+            [|
+              do
+                  -- force creation of datatype
+                  _ <- typeRep (undefined :: Proxy $(TH.conT tyName))
+
+                  -- get the right constructor
+                  maybeConstructors <- lookupConstructors () tyName
+                  constrs <- case maybeConstructors of
+                      Nothing -> die $ "Constructors not created for " ++ show tyName
+                      Just cs -> pure cs
+                  let constr = constrs !! index
+
+                  -- need to instantiate it to the right types and then lift all the arguments and apply it to them
+                  types <- sequence $(TH.listE typeExprs)
+                  lifts <- sequence $(TH.listE liftExprs)
+
+                  pure $ mkIterApp () (mkIterInst () constr types) lifts
+            |]
+        TH.clause [pat] (TH.normalB $ pure expr) []
+
+makeLiftPir :: TH.Name -> TH.Q [TH.Dec]
+makeLiftPir name = do
```

Comment:
Is it possible to check at this point that the required extensions are enabled, and give a nice error message if they aren't? Esp.

> - `ScopedTypeVariables` (*terrible* error)

Reply:
Hm, actually, maybe we can do that.

Reply:
Done!

Reply:
To be clear, I only did this for `ScopedTypeVariables`, the others you get a sensible error from GHC.